pkg/storage/store.go

```diff
@@ -461,15 +460,34 @@ func (s *LokiStore) SelectSeries(ctx context.Context, req logql.SelectLogParams)
 	return result, nil
 }
 
+func extractLineFilters(p *plan.QueryPlan) []syntax.LineFilter {
+	lineFilters := make([]syntax.LineFilter, 0)
+	visitor := &syntax.DepthFirstTraversal{
+		VisitLineFilterFn: func(v syntax.RootVisitor, e *syntax.LineFilterExpr) {
+			if e.Left != nil {
+				e.Left.Accept(v)
+			}
+			if e.Or != nil {
+				e.Or.Accept(v)
+			}
+			lineFilters = append(lineFilters, e.LineFilter)
```

Comment:
Neat! I would have created a new visitor that embeds the default visitor but this is much cleaner and easier to understand.